### PR TITLE
Feature/failure

### PR DESCRIPTION
**Pull request by hoodie**

### CODE REVIEW

**Comment by hoodie** [comment]
@mkroehnert if you want to help I'd appreciate of you could review this

File: examples/mac_app_id.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        .icon("Safari")
+        .show()
+        .map_err(|f| format!("{}", f))?;
+
+    Ok(())
+}
+
+#[cfg(all(unix, not(target_os = "macos")))]
+fn main() {
+    println!("this is a mac only feature")
```
Comment (mkroehnert) [inline examples/mac_app_id.rs]:
```suggestion
    println!("get_bundle_identifier_or_default() and set_application_id() are macOS only features")
```

File: examples/mac_app_id.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#[cfg(target_os = "macos")]
+fn main() -> Result<(), String> {
+    use notify_rust::{Notification, get_bundle_identifier_or_default, set_application};
+
+    let safari_id = get_bundle_identifier_or_default("Safari");
+    set_application(&safari_id)
+        .map_err(|f| format!("{}", f))?;
+
+    set_application(&safari_id)
+        .map_err(|f| format!("{}", f))?;
```
Comment (mkroehnert) [inline examples/mac_app_id.rs]: Why is this called twice?
A comment would help in this case.

File: examples/mac.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        .appname("Safari")
+        .icon("Safari")
+        .show()?;
+
+    Ok(())
+}
+
+#[cfg(all(unix, not(target_os = "macos")))]
+fn main() {
+    println!("this is a mac only feature")
```
Comment (mkroehnert) [inline examples/mac.rs]: Same here:
```suggestion
    println!("get_bundle_identifier_or_default() and set_application_id() are macOS only features")
```

File: examples/mac.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+
+extern crate notify_rust;
+
+#[cfg(target_os = "macos")]
+fn main() -> Result<(), Box<std::error::Error>> {
+    use notify_rust::{Notification, get_bundle_identifier_or_default, set_application};
+
+    let safari_id = get_bundle_identifier_or_default("Safari");
+    set_application(&safari_id)?;
+    set_application(&safari_id)?;
```
Comment (mkroehnert) [inline examples/mac.rs]: Why is this method called twice?
If there is a specific reason, a comment would help.

File: README.md [hunk 5]
```diff
@@ -62,6 +61,8 @@ Please see the [documentation](https://docs.rs/crate/notify-rust/) for current e
 ### Commandline tool
 Checkout [toastify](https://github.com/hoodie/toastify), it exposes most of the functionality of the lib to the commandline.
 
-### Platform support
+### macOS support
 
-This library shines on linux and bsd, which is it's original target platform. Lately it gained support for macOS, however this only includes a small subset of the current functionality, since [`NSNotification`](https://developer.apple.com/reference/foundation/nsnotification)s don't have as many features. Please refer to the You are a versed macOS UI developer with mad Objective-C skillz? <abbr title="pull request sil vous plait">PRSV</abbr>.
+This library shines on linux and bsd, which is it's original target platform.
+Lately it gained support for macOS thanks to [mac-notification-sys](https://crates.io/crates/mac-notification-sys).
+However this only includes a small subset of the current functionality, since [`NSNotification`](https://developer.apple.com/reference/foundation/nsnotification)s don't have as many features. Please refer to the You are a versed macOS UI developer with mad Objective-C skillz? <abbr title="pull request sil vous plait">PRSV</abbr>.
```
Comment (mkroehnert) [inline README.md]: Something is missing at the end of the sentence.

Comment (hoodie) [reply README.md]: I was kinda hoping for some feedback on the error module 😁

Comment (mkroehnert) [reply README.md]: @hoodie I'll read up on the failure crate tomorrow. But as already said, I haven't used it so far.

File: src/error.rs [hunk 6]
```diff
@@ -1,27 +1,37 @@
...
-        Error, ErrorKind, ResultExt, Result;
-    }
+    #[fail(display="Parsing Error")]
+    Parse(#[cause] num::ParseIntError),
+
+    #[fail(display="The running server supplied an unknown version: {}", _0 )]
+    SpecVersion(String),
+
+    #[fail(display="ParseError: {}", error)]
+    ParseError{ #[cause] error: ::std::num::ParseIntError },
```
Comment (mkroehnert) [inline src/error.rs]: Hi @hoodie, I have been reading up on the failure crate, and the changes here look fine.

There is only one thing I do not yet understand. Why is there `Parse` and `ParseError`, when both store a `::std::num::ParseIntError`? To me, it looks like the later one would be sufficient and should be returned by `From<num::ParseIntError>`.

**Comment by mkroehnert** [comment]
Compilation of the branch on Linux worked for me and all unittests passed.

**Pull request closed without merging**